src/gpu/intel/jit/gemm/gen_gemm_kernel.cpp

```diff
@@ -527,55 +527,70 @@ status_t gen_gemm_nocopy_kernel_desc_t::select_kernel(compute::gpu_arch_t arch,
         if (!has_mode) return;
         auto &def = base.selector.precisions;
         if (match(problem_.Ta)) {
-            match_params.push_back(base);
+            if (optional) {
+                match_params.push_back(base);
+                match_params.back().selector.precisions[1] = def[1];
+            }
             match_params.back().selector.precisions[0] = match(problem_.Ta);
-            match_params.back().selector.precisions[1] = def[1];
         }
         if (match(problem_.Tb)) {
-            match_params.push_back(base);
-            match_params.back().selector.precisions[0] = def[0];
+            if (optional) {
+                match_params.push_back(base);
+                match_params.back().selector.precisions[0] = def[0];
+            }
             match_params.back().selector.precisions[1] = match(problem_.Tb);
         }
         if (match(problem_.Ta) && match(problem_.Tb)) {
-            match_params.push_back(base);
+            if (optional) match_params.push_back(base);
             match_params.back().selector.precisions[0] = match(problem_.Ta);
             match_params.back().selector.precisions[1] = match(problem_.Tb);
         }
     };
 
-    add_mode_matches(fpmath_tf32, [](Type dt) -> const char * {
-        if (dt == Type::f32) { return "T"; }
-        return nullptr;
-    });
-
-    add_mode_matches(fpmath_bf16, [](Type dt) -> const char * {
-        if (dt == Type::f32) { return "[SB]"; }
-        if (dt.isInt8() || dt.isInt4()) return "[OB]";
-        if (dt.isF8()) return "B";
-        return nullptr;
-    });
-
-    add_mode_matches(fpmath_f16, [](Type dt) -> const char * {
-        if (dt == Type::f32) { return "[SH]"; }
-        if (dt.isInt8() || dt.isInt4()) return "[OH]";
-        if (dt.isF8()) return "H";
-        return nullptr;
-    });
-
-    add_mode_matches(!(fpmath_f16 || fpmath_bf16), [](Type dt) -> const char * {
-        if (dt.isInt4()) return "[FO]";
-        return nullptr;
-    });
+    add_mode_matches(
+            fpmath_tf32, /*optional=*/true, [](Type dt) -> const char * {
+                if (dt == Type::f32) { return "T"; }
+                return nullptr;
+            });
+
+    add_mode_matches(
+            fpmath_bf16, /*optional=*/true, [](Type dt) -> const char * {
+                if (dt == Type::f32) { return "[SB]"; }
+                if (dt.isInt8() || dt.isInt4()) return "[OB]";
+                if (dt.isF8()) return "B";
+                return nullptr;
+            });
+
+    add_mode_matches(
+            fpmath_f16, /*optional=*/true, [](Type dt) -> const char * {
+                if (dt == Type::f32) { return "[SH]"; }
+                if (dt.isInt8() || dt.isInt4()) return "[OH]";
+                if (dt.isF8()) return "H";
+                return nullptr;
+            });
+
+    add_mode_matches(!(fpmath_f16 || fpmath_bf16), /*optional=*/false,
+            [](Type dt) -> const char * {
+                if (dt.isInt4()) return "[FO]";
```

Comment:
I'm not sure this part is 100% correct. We have two kinds of x8:x4:f16 kernels in the catalog: `FOS` and `[FO]OS`. We can always use the `[FO]OS` ones (upconverting int4->int8 prior to dpas). However, if there are no wei zp then we can pass the int4 wei data directly to dpas -- `FOS` -- and that can be faster. Here, then, if I'm reading the patch correctly, you want optional to be determined by the presence of wei zp.